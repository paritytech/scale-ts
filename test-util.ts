/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
