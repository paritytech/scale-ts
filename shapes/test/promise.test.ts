import * as $ from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

testShape($.promise($.u8), {
  0: () => prom(0),
  255: () => prom(255),
}, true)

testShape($.array($.promise($.u8)), {
  empty: () => [],
  single: () => [prom(0)],
  sequential: () => Array.from({ length: 256 }, (_, i) => prom(i)),
  times13: () => Array.from({ length: 256 }, (_, i) => prom(i * 13 % 256)),
}, true)

testShape($.promise($.array($.promise($.compact($.u32)))), {
  times13: () => Promise.resolve(Array.from({ length: 256 }, (_, i) => prom(i * 13 % 256))),
}, true)

testInvalid($.promise($.u8), [0])

function prom(n: number): Promise<number> {
  return new Promise((r) => setTimeout(() => r(n), n * 10))
}
