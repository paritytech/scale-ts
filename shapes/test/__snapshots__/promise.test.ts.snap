export const snapshot = {};

snapshot[`\$.promise(\$.u8) 0 1`] = `00`;

snapshot[`\$.promise(\$.u8) 255 1`] = `ff`;

snapshot[`\$.array(\$.promise(\$.u8)) empty 1`] = `00`;

snapshot[`\$.array(\$.promise(\$.u8)) single 1`] = `
04
00
`;

snapshot[`\$.array(\$.promise(\$.u8)) sequential 1`] = `
01
04
00
01
02
03
04
05
06
07
08
09
0a
0b
0c
0d
0e
0f
10
11
12
13
14
15
16
17
18
19
1a
1b
1c
1d
1e
1f
20
21
22
23
24
25
26
27
28
29
2a
2b
2c
2d
2e
2f
30
31
32
33
34
35
36
37
38
39
3a
3b
3c
3d
3e
3f
40
41
42
43
44
45
46
47
48
49
4a
4b
4c
4d
4e
4f
50
51
52
53
54
55
56
57
58
59
5a
5b
5c
5d
5e
5f
60
61
62
63
64
65
66
67
68
69
6a
6b
6c
6d
6e
6f
70
71
72
73
74
75
76
77
78
79
7a
7b
7c
7d
7e
7f
80
81
82
83
84
85
86
87
88
89
8a
8b
8c
8d
8e
8f
90
91
92
93
94
95
96
97
98
99
9a
9b
9c
9d
9e
9f
a0
a1
a2
a3
a4
a5
a6
a7
a8
a9
aa
ab
ac
ad
ae
af
b0
b1
b2
b3
b4
b5
b6
b7
b8
b9
ba
bb
bc
bd
be
bf
c0
c1
c2
c3
c4
c5
c6
c7
c8
c9
ca
cb
cc
cd
ce
cf
d0
d1
d2
d3
d4
d5
d6
d7
d8
d9
da
db
dc
dd
de
df
e0
e1
e2
e3
e4
e5
e6
e7
e8
e9
ea
eb
ec
ed
ee
ef
f0
f1
f2
f3
f4
f5
f6
f7
f8
f9
fa
fb
fc
fd
fe
ff
`;

snapshot[`\$.array(\$.promise(\$.u8)) times13 1`] = `
01
04
00
0d
1a
27
34
41
4e
5b
68
75
82
8f
9c
a9
b6
c3
d0
dd
ea
f7
04
11
1e
2b
38
45
52
5f
6c
79
86
93
a0
ad
ba
c7
d4
e1
ee
fb
08
15
22
2f
3c
49
56
63
70
7d
8a
97
a4
b1
be
cb
d8
e5
f2
ff
0c
19
26
33
40
4d
5a
67
74
81
8e
9b
a8
b5
c2
cf
dc
e9
f6
03
10
1d
2a
37
44
51
5e
6b
78
85
92
9f
ac
b9
c6
d3
e0
ed
fa
07
14
21
2e
3b
48
55
62
6f
7c
89
96
a3
b0
bd
ca
d7
e4
f1
fe
0b
18
25
32
3f
4c
59
66
73
80
8d
9a
a7
b4
c1
ce
db
e8
f5
02
0f
1c
29
36
43
50
5d
6a
77
84
91
9e
ab
b8
c5
d2
df
ec
f9
06
13
20
2d
3a
47
54
61
6e
7b
88
95
a2
af
bc
c9
d6
e3
f0
fd
0a
17
24
31
3e
4b
58
65
72
7f
8c
99
a6
b3
c0
cd
da
e7
f4
01
0e
1b
28
35
42
4f
5c
69
76
83
90
9d
aa
b7
c4
d1
de
eb
f8
05
12
1f
2c
39
46
53
60
6d
7a
87
94
a1
ae
bb
c8
d5
e2
ef
fc
09
16
23
30
3d
4a
57
64
71
7e
8b
98
a5
b2
bf
cc
d9
e6
f3
`;

snapshot[`\$.promise(\$.array(\$.promise(\$.compact(\$.u32)))) times13 1`] = `
01
04
00
34
68
9c
d0
05
01
39
01
6d
01
a1
01
d5
01
09
02
3d
02
71
02
a5
02
d9
02
0d
03
41
03
75
03
a9
03
dd
03
10
44
78
ac
e0
15
01
49
01
7d
01
b1
01
e5
01
19
02
4d
02
81
02
b5
02
e9
02
1d
03
51
03
85
03
b9
03
ed
03
20
54
88
bc
f0
25
01
59
01
8d
01
c1
01
f5
01
29
02
5d
02
91
02
c5
02
f9
02
2d
03
61
03
95
03
c9
03
fd
03
30
64
98
cc
01
01
35
01
69
01
9d
01
d1
01
05
02
39
02
6d
02
a1
02
d5
02
09
03
3d
03
71
03
a5
03
d9
03
0c
40
74
a8
dc
11
01
45
01
79
01
ad
01
e1
01
15
02
49
02
7d
02
b1
02
e5
02
19
03
4d
03
81
03
b5
03
e9
03
1c
50
84
b8
ec
21
01
55
01
89
01
bd
01
f1
01
25
02
59
02
8d
02
c1
02
f5
02
29
03
5d
03
91
03
c5
03
f9
03
2c
60
94
c8
fc
31
01
65
01
99
01
cd
01
01
02
35
02
69
02
9d
02
d1
02
05
03
39
03
6d
03
a1
03
d5
03
08
3c
70
a4
d8
0d
01
41
01
75
01
a9
01
dd
01
11
02
45
02
79
02
ad
02
e1
02
15
03
49
03
7d
03
b1
03
e5
03
18
4c
80
b4
e8
1d
01
51
01
85
01
b9
01
ed
01
21
02
55
02
89
02
bd
02
f1
02
25
03
59
03
8d
03
c1
03
f5
03
28
5c
90
c4
f8
2d
01
61
01
95
01
c9
01
fd
01
31
02
65
02
99
02
cd
02
01
03
35
03
69
03
9d
03
d1
03
04
38
6c
a0
d4
09
01
3d
01
71
01
a5
01
d9
01
0d
02
41
02
75
02
a9
02
dd
02
11
03
45
03
79
03
ad
03
e1
03
14
48
7c
b0
e4
19
01
4d
01
81
01
b5
01
e9
01
1d
02
51
02
85
02
b9
02
ed
02
21
03
55
03
89
03
bd
03
f1
03
24
58
8c
c0
f4
29
01
5d
01
91
01
c5
01
f9
01
2d
02
61
02
95
02
c9
02
fd
02
31
03
65
03
99
03
cd
03
`;

snapshot[`\$.promise(\$.u8) invalid 0 1`] = `ShapeAssertError: !(value instanceof Promise)`;
